### PR TITLE
MDEV-14787 CREATE does not allow tables with versioning columns only, but ALTER does

### PR DESCRIPTION
**Pull request by kevgs**

### CODE REVIEW

**Review by midenok — CHANGES_REQUESTED** [review]
Don't do any additional loops. Use `done_start`, `done_end` flags and comparison of elements count between `drop_list` and `share->fields`.

**Comment by kevgs** [comment]
Updated.

**Review by midenok — CHANGES_REQUESTED** [review]
No, you didn't do the full fix. Either comment why you didn't do it or fix. Don't make me reject your PRs multiple times.

**Comment by kevgs** [comment]
At the point of
```diff
diff --git a/sql/handler.cc b/sql/handler.cc
index 60909ba2299..e4f1d480755 100644
--- a/sql/handler.cc
+++ b/sql/handler.cc
@@ -7342,6 +7342,12 @@ bool Vers_parse_info::fix_alter_info(THD *thd, Alter_info *alter_info,
         if (done_start && done_end)
           break;
       }
+
+      if (share->fields - alter_info->drop_list.elements <= 2)
+      {
+        my_error(ER_VERS_TABLE_MUST_HAVE_COLUMNS, MYF(0), table_name);
+        return true;
+      }
     }
 
     return false;
```
`done_start`, `done_end` are not needed because it was removed from `alter_info->drop_list`. And we just compare element counts. Is it ok?

**Review by midenok — APPROVED** [review]

File: sql/handler.cc [hunk 1]
```diff
@@ -7342,6 +7342,12 @@ bool Vers_parse_info::fix_alter_info(THD *thd, Alter_info *alter_info,
         if (done_start && done_end)
           break;
       }
+
+      if (share->fields - alter_info->drop_list.elements <= 2)
```
Comment (midenok) [inline sql/handler.cc]: We have constant `VERSIONING_FIELDS` for number of system fields.

**Comment by kevgs** [comment]
This PR doesn't handle duplicates in `drop_list` and thus breaks error messages.
Before patch:
```sql
create or replace table t (a int, b int) with system versioning;

MariaDB [test]> alter table t drop b, drop b;
ERROR 1091 (42000): Can't DROP COLUMN `b`; check that it exists
```

After patch:
```sql
MariaDB [test]> create or replace table t (a int, b int) with system versioning;

MariaDB [test]> alter table t drop b, drop b;
ERROR 4128 (HY000): Table `t` must have at least one temporal column
```

First version of PR doesn't have such flaw. Maybe improve it somehow or take as is?

**Comment by midenok** [comment]
I don't see much flaw here. It prohibited first `drop b` and this is right.

**Comment by midenok** [comment]
This check is done here:
```c++
#0  my_error (nr=1091, MyFlags=0) at /home/midenok/src/mariadb/trunk/src/mysys/my_error.c:113
#1  0x000000000084675f in mysql_prepare_alter_table (thd=0x7fff80000d50, table=0x7fff80098960, create_info=0x7fffe52aa418, alter_info=0x7fffe52aa370, alter_ctx=0x7fffe52a9110) at /home/midenok/src/mariadb/trunk/src/sql/sql_table.cc:8356
#2  0x0000000000848f27 in mysql_alter_table (thd=0x7fff80000d50, new_db=0x7fff80013f90 "test", new_name=0x0, create_info=0x7fffe52aa418, table_list=0x7fff80013960, alter_info=0x7fffe52aa370, order_num=0, order=0x0, ignore=false) at /home/midenok/src/mariadb/trunk/src/sql/sql_table.cc:9330
#3  0x00000000008f4d08 in Sql_cmd_alter_table::execute (this=0x7fff80013fe8, thd=0x7fff80000d50) at /home/midenok/src/mariadb/trunk/src/sql/sql_alter.cc:325
#4  0x00000000007529b4 in mysql_execute_command (thd=0x7fff80000d50) at /home/midenok/src/mariadb/trunk/src/sql/sql_parse.cc:6261
#5  0x00000000007446ef in mysql_parse (thd=0x7fff80000d50, rawbuf=0x7fff80013878 "alter table t drop b, drop b", length=28, parser_state=0x7fffe52ad650, is_com_multi=false, is_next_command=false) at /home/midenok/src/mariadb/trunk/src/sql/sql_parse.cc:7991
```
Move your check after it with proper adjustments. Please, fix comment about constant.

**Comment by kevgs** [comment]
Thanks. Updated.

**Review by midenok — APPROVED** [review]
Please, merge.

**Pull request merged**